src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+
+  private val beamConfig: BeamConfig = beamScenario.beamConfig
+  private val cnmConfig = beamConfig.beam.agentsim.chargingNetworkManager
+
+  private val chargingNetworkMap = TrieMap.empty[String, ChargingNetwork]
+  private val chargingZoneList: QuadTree[ChargingZone] = loadChargingZones(beamServices)
+  private lazy val sitePowerManager = new SitePowerManager(chargingNetworkMap, beamServices)
+  private lazy val powerController = new PowerController(chargingNetworkMap, beamConfig)
+  private val endOfSimulationTime: Int = DateUtils.getEndOfTime(beamConfig)
+
+  import powerController._
+  import sitePowerManager._
+
+  private def currentTimeBin(tick: Int): Int = cnmConfig.timeStepInSeconds * (tick / cnmConfig.timeStepInSeconds)
+  private def nextTimeBin(tick: Int): Int = currentTimeBin(tick) + cnmConfig.timeStepInSeconds
+
+  override def receive: Receive = {
+    case TriggerWithId(InitializeTrigger(_), triggerId) =>
+      import scala.concurrent.{ExecutionContext, Future}
+      implicit val timeout: Timeout = Timeout(50000, TimeUnit.SECONDS)
+      implicit val executionContext: ExecutionContext = context.dispatcher
+      chargingNetworkMap.put(defaultVehicleManager, new ChargingNetwork(defaultVehicleManager, chargingZoneList))
+      Future(scheduler ? ScheduleTrigger(PlanningTimeOutTrigger(0), self))
+        .map(_ => CompletionNotice(triggerId, Vector()))
+        .pipeTo(sender())
+
+    case TriggerWithId(PlanningTimeOutTrigger(timeBin), triggerId) =>
+      log.debug(s"Planning energy dispatch for vehicles currently connected to a charging point, at t=$timeBin")
+      import scala.concurrent.ExecutionContext.Implicits.global
+      import scala.concurrent.duration._
+      import scala.concurrent.{Await, Future, TimeoutException}
+      val estimatedLoad = requiredPowerInKWOverNextPlanningHorizon(timeBin)
+      log.debug("Total Load estimated is {} at tick {}", estimatedLoad.values.sum, timeBin)
+
+      // obtaining physical bounds
+      val physicalBounds = obtainPowerPhysicalBounds(timeBin, Some(estimatedLoad))
+
+      val futures = Future
+        .sequence(chargingNetworkMap.flatMap {
+          case (_, chargingNetwork) =>
+            chargingNetwork.lookupConnectedVehicles.map {
+              case (_, chargingVehicle @ ChargingVehicle(vehicle, vehicleManager, _, _, _, _)) =>
+                Future {
+                  // Refuel
+                  handleRefueling(chargingVehicle)
+                  // Calculate the energy to charge each vehicle connected to the a charging station
+                  val (chargingDuration, energyToCharge) =
+                    dispatchEnergy(cnmConfig.timeStepInSeconds, chargingVehicle, physicalBounds)
+                  // update charging vehicle with dispatched energy and schedule ChargingTimeOutScheduleTrigger
+                  chargingVehicle.processChargingCycle(timeBin, energyToCharge, chargingDuration).flatMap {
+                    case cycle if cycle.duration == 0 =>
```

Comment:
This should be changed to a `def` that names what this `if` means? I still am not 100%....following the logic it seems that the current impl is that the duration is really just the `cnmConfig.timeStepInSeconds` (dispatch returns the duration as the timestep, then process passes the duration back out (unless it is the same starttime))..

Reply:
Actually...even the further cycles are confusing me...Will ping offline to discuss the final logic a bit more.

Reply:
good idea